### ISSUE TITLE
Filename not properly interpreted by TFTP

### ISSUE THREAD

**Issue by BloodBlight** [netbootxyz/netboot.xyz#1198]
**Describe the bug**
This is related to this discussion from a while back:
https://github.com/netbootxyz/netboot.xyz/discussions/1133

I have an older client is sending a file name to the TFTP server, but is not being properly decoded into Unicode.  I am not sure if the TFTP server is from another project (and this should be posted there), so please feel free to re-direct me and I will post it there.

What I see in the logs:
```2022-07-08 15:28:56 notice in.tftpd[88]: RRQ from 192.168.10.206 filename netboot.xyz.kpxe�����```

The last characters (�) is (U+FFFD), a Unicode character for "Replacement Character".  As it is at the end of the string, there could be a few causes, but without knowing the code...

**To Reproduce**
Attempt to boot from PXE.

**Expected behavior**
Should download the file and start booting.

**Additional context**
This is an old board, but a good one (kinda like a raspberry Pi, years before they were a thing.  I can't find the exact model, but it is basically a Jetway J7F4K1G5DS-LF, just a bit slower (1.2GHz).

**Comment by MartinLoeper** [comment]
I am noticing the same for a Lenovo T460 laptop. I am getting the following hex dump on the wire: 6e 65 74 62 6f 6f 74 2e 78 79 7a 2e 65 66 69 ff 00. The ff byte before the nullbyte should not be there. Is this a bug in the uefi firmware or am I missing something?

**Comment by antonym** [comment]
Not really sure if there's anything on my side I can do about it, using Proxmox VMs to my docker container running tftp it seems to pull up without issues:

```
2023-10-18 03:54:57 notice in.tftpd[85]: RRQ from 10.0.100.155 filename netboot.xyz.efi
```

Wonder if whatever is serving up the filename for DHCP is putting some characters on the file name? Where are you setting next-server and setting the filename from? Is there a common router that may be appending some characters to the filename?

I think I've seen some other scenarios where people have had extra characters added to their filename, it could be a bug in a common router you might be using as well:

https://www.reddit.com/r/linuxquestions/comments/4rwesw/tftpboot_invalid_character_in_requested_filename/

**Comment by BloodBlight** [comment]
I can have two computers sitting next to each other, one works, the other doesn't.  Same target.  I will verify that though (might not be today) to make sure I am not full of it.  So I am fairly sure it is something about the way the client is making the request.

DHCP Is being handled by a UniFi USG 3.  Fairly basic config:
https://imgur.com/a/ZnOfpZw

No idea why Imgur is tagging it as mature...  Ignore.

**Comment by megapearl** [comment]
Having the same problem here.

Client:

Intel UNDI, PXE-2.1 (build 082)
Copyright (C) 1997-2000 Intel Corporation

This Product is covered by one or more of the following patents:
US5,307,459, US5,434,872, US5,732,094, US6,570,884, US6,115,776 and
US6,327,625

Realtek PCI Express Gigabit Ethernet Controller Series v2.26 (090219)

CLIENT MAC ADDR: 90 FB A6 41 09 ED CUID: C996DEBD-1DF6-1E31-E57C-90FBA64109ED
CLIENT IP: 10.0.0.100
MASK: 255.255.255.0 DHCP IP: 10.0.0.1
CATEWAY IP: 10.0.0.1

TFTP .
PXE-T01: File not found
PXE-E3B: TFTP Error - File Not found
PSE-MOF: Exiting PXE ROM.
Reboot and Select proper Boot device or Insert Boot Media in selected Boot device and press a key

Server:
![image](https://github.com/netbootxyz/netboot.xyz/assets/3420660/06b2ff24-be54-4498-afc3-2278001f2b9d)

**Comment by BloodBlight** [comment]
I am tempted to do a packet capture...

I have had multiple older systems have this issue now...

**Comment by rudiservo** [comment]
The issue is in the DHCP server passing the information to old nics
In my case if you have this option

`option bootfile-name "netboot.xyz.kpxe";`
 
 the old nic will put some special chars on the end of it.
 
 if you have the config has specified in the [docs](https://netboot.xyz/docs/docker)
 
 you have something like this in the dhcp server
 
 ```
 if exists user-class and ( option user-class = "iPXE" ) {
    filename "http://boot.netboot.xyz/menu.ipxe";
  } elsif option arch = encode-int ( 16, 16 ) {
    filename "http://boot.netboot.xyz/ipxe/netboot.xyz.efi";
    option vendor-class-identifier "HTTPClient";
  } elsif option arch = 00:07 {
    filename "netboot.xyz.efi";
  } else {
    filename "netboot.xyz.kpxe";
  }
  ```

so try to remove option boot-filename if you NIC is to old.

**Comment by antonym** [comment]
Let me know if dropping the option helps and we can get the knowledge base updated.

**Comment by BloodBlight** [comment]
I will test later tonight and let you know.

**Comment by BloodBlight** [comment]
So, I didn't fully read through the instructions at work the other day...  I am using a UniFi device for my DHCP and can't deploy a config like that.  I only get the option to include a host and a file name:
https://imgur.com/a/aYicWfO

Suggestions other than a rip and replace?

**Comment by rudiservo** [comment]
@BloodBlight does it work without TFTP server?
Unifi are great, except when you need to dig deep.
Can you ssh into the router?

**Comment by BloodBlight** [comment]
Yeeeeesss...  :)  But I am not really comfortable editing their their config without the GUI's consent.  That has a bad tendency to break things.

If they don't have the options, then they don't and I can work around that.  Maybe have a "deployment" network that has it's own server.

I can test that with these boxes, but it will take a bit more time.  Too many plates and not enough spoons to handle that one right away.

**Comment by rudiservo** [comment]
I feel your pain.
Yes test it with the boxes
because if it does you can put a support request for unifi and hopefully they will add more options to fix it in the router GUI.

**Comment by megapearl** [comment]
> The issue is in the DHCP server passing the information to old nics In my case if you have this option
> 
> `option bootfile-name "netboot.xyz.kpxe";`
> 
> the old nic will put some special chars on the end of it.
> 
> if you have the config has specified in the [docs](https://netboot.xyz/docs/docker)
> 
> you have something like this in the dhcp server
> 
> ```
> if exists user-class and ( option user-class = "iPXE" ) {
>    filename "http://boot.netboot.xyz/menu.ipxe";
>  } elsif option arch = encode-int ( 16, 16 ) {
>    filename "http://boot.netboot.xyz/ipxe/netboot.xyz.efi";
>    option vendor-class-identifier "HTTPClient";
>  } elsif option arch = 00:07 {
>    filename "netboot.xyz.efi";
>  } else {
>    filename "netboot.xyz.kpxe";
>  }
> ```
> 
> so try to remove option boot-filename if you NIC is to old.

Tried to remove the boot-filename parameter, but then the iPXE client cannot find a bootfile and is not booting at all.
I got it working by replacing the DHCP server in pfSense 2.7.2-RELEASE from the new Kea one to the old isc-dhcp deprecated one.

![image](https://github.com/netbootxyz/netboot.xyz/assets/3420660/3c0a9afe-0fac-4d99-b57e-41f48b3d785e)

Now booting old bios clients and new uefi clients.

**Comment by rudiservo** [comment]
@megapearl I have OpnSense so the way PfSense and OpnSense generate the config file may be different, best way to check this is by going strait into ssh and cat the config file to check if the option is there.

Anyway it's good to know that at least in Kea generated file that issue is sort of mitigated.

IMO advisory should be added to docs about this issue and that it may differ from router firmware or software.

**Comment by eladiogalvez** [comment]
Same issue with pfsense and old nic.
Fixed using deprecated dhcp server as @megapearl commented.

**Comment by gbuloz** [comment]
Same issue with kea 2.6.2 on Fedora 41 when booting a 10 years old Kontron VX3040 Intel CPU board. After some debug with tcpdump I noticed an extra FF at the end of the filename requested to the TFTP server. As a workaround I've renamed the boot files on the server with an extra FF at the end of their names using : mv name.efi $(echo -en 'name.efi\xFF')

**Comment by BloodBlight** [comment]
Oh, that's a great little hack!  I can even script out that creation.  Thank you for sharing!

**Comment by Koli0842** [comment]
I have also encountered this (Ubiquiti USG, and an old Z77X-UD3H with an Atheros NIC), the above hack worked. Instead of moving, I created symlinks for all bootable entities, it still works

**Comment by BloodBlight** [comment]
As this is NOT a bug with netbootxyz, but rather an issue with with DHCP, I am open to closing this ticket.  Maybe taking the notes from it and this "hack" and putting it into the docs?

**Comment by Koli0842** [comment]
Sounds good to me, took some digging to find this.

**Comment by bwann** [comment]
I did a bunch of digging into this problem and reading of UEFI specs a while back and I'm pretty sure what is happening is UEFI bugs around line terminations, at least when using UEFI PXE on IPv4. This comes down to using the fixed length "file" field in the DHCP header vs using Option 67 for providing a boot filename, and how different UEFI implementations handle line and field terminations in DHCPv4 packets.

RFC 2132 does not guarantee the filename in option 67 is null terminated. In the UEFI 2.6 spec DHCP options are treated as  variable length fields in the form of "option code + length of option data + option data" but the spec does not address line terminations at all, that's left as an implementation detail. So some handle the filename in option 67 correctly and some don't.

For example for several of my SuperMicro boards given only option 67 (no higher numbered options) as "testfilename", the packet will contain 0x48 (decimal 67), 0x0C (length 12), the filename, and then 0xff at the end of the packet. They should read 12 bytes and stop, but what they do is keep reading data in as the filename right to the end of the packet as if they're expecting a null terminated string and not a variable-length-field. That's what gets the junk added to the end of TFTP filenames.  I demonstrated this by adding some data to option 68 so it follows next in the same packet and their UEFI PXE code kept reading in the data in option 68 as part of the TFTP filename!

The way I see to fix this is not to use option 67 when providing a boot filename, but the actual 'file' field in the DHCP header. At least in Kea this is done as providing "boot-file-name" in the top level Dhcp4 (or class) config, and not in "option-data". Or pray and complain to the vendor they fix their UEFI.

Interestingly DHCPv6 doesn't have this problem because it's all variable-length-fields. The Tianocore EDK II reference UEFI implementation also goes out of its way to call out the ambiguous line termination with Option 67 and tries to do the right thing.  I have a writeup a did that contains code snippets and examples: https://binaryfury.wann.net/2024/12/pxe-two-tftp-filenames-one-dhcp-offer-plus-weird-filenames/

**Comment by antonym** [comment]
@claude can you summarize this for a kb article for https://github.com/netbootxyz/netboot.xyz-docs
